### DISCUSSION TITLE
azure devops authentication with service principal

### DISCUSSION THREAD

**Discussion by odombrovska** [fluxcd/flux2#3765] [Q&A]
i am trying to use service principal that has access to azure devops repo for authentication, providing its appId and clientSecret. however, i was only able to authenticate with PAT. is authentication with service principal currently not supported? only ssh and PAT are supported?

**Comment by rsitrifork** [comment]
Yes, the service principal authentication requires the source controller to pull a token from the AAD, the credentials cannot be used directly. 

But I would love for this feature to be implemented, especially the managed identity version, so we can get away from having credentials on the cluster

**Comment by kingdonb** [comment]
We've not seen this capability in Azure docs, is it possible to authenticate directly with service principal? I think we found that PAT was the only available path, at least at the time that it was implemented here. If you can provide a link to a doc that contradicts this, we'll have a look.

Currently there is no autologin provider for Git support in Flux today. We only have support for autologin for buckets, sources like OCI and for ImageRepository resources. So it would be a new feature, adding context-based login for a Git provider.

You can see all of the Flux sources that have implemented a context-driven autologin feature like this here:

https://fluxcd.io/flux/security/contextual-authorization/

**Reply by rsitrifork** [reply]
It is quite new: https://devblogs.microsoft.com/devops/introducing-service-principal-and-managed-identity-support-on-azure-devops/

**Reply by kingdonb** [reply]
Ah, thanks for that – looks like it landed in public preview just a bit over 1 month ago. @darkowlzz FYI

**Reply by m1rl0k** [reply]
trying to do this for nuget feed myself i need in github actions; instead of a pat to get the nuget feed from azure devops
